pire/scanners/multi.h

```diff
@@ -1041,21 +1050,191 @@ class ScannerGlueTask: public ScannerGlueCommon<Scanner> {
 		Sc().BuildShortcuts();
 		return Sc();
 	}
+
+	template<class Iter, class OutIter>
+	static OutIter Shift(ypair<Iter, Iter> range, size_t shift, OutIter out)
+	{
+		for (; range.first != range.second; ++range.first, ++out)
+			*out = *range.first + shift;
+		return out;
+	}
 	
 private:    
 	template<class Iter>
 	size_t RangeLen(ypair<Iter, Iter> range) const
 	{
 		return std::distance(range.first, range.second);
 	}
+};
 
-	template<class Iter, class OutIter>
-	OutIter Shift(ypair<Iter, Iter> range, size_t shift, OutIter out) const
+template <typename Scanner>
+class ScannerStateRegexpHash {
+public:
+	ScannerStateRegexpHash(const Scanner& scanner)
+		: m_scanner(scanner)
+	{}
+
+	size_t operator()(const size_t& state) const {
+		const auto regexps = m_scanner.AcceptedRegexps(state);
+		size_t result = 0;
+		for (auto iter = regexps.first; iter != regexps.second; ++iter) {
+			result ^= std::hash<size_t>{}(*iter);
```

Comment:
IMHO
Maybe it would be better to choose a non-commutative scheme of hash combining.
For example borrow it from `boost::hash_combine`.